[Feature] Add PyMuPDF extraction integration via extralit-hf-space service

priyankeshh opened the pull request:
## Description

This PR adds PyMuPDF extraction functionality by integrating extralit-server document upload job with the extralit-hf-space microservice. The integration triggers PDF extraction during document upload and merges the results with existing document metadata.

## Related Tickets & Documents

Closes # <!-- Add issue number if available -->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Steps to QA

1. Start the extralit-hf-space service:
   ```bash
   cd extralit-hf-space
   python -m uvicorn src.app:app --host 0.0.0.0 --port 7860 --reload
   ```

2. Set the environment variable for hf-space URL:
   ```bash
   export HF_SPACE_URL=http://localhost:7860
   ```

3. Test the integration with provided test scripts:
   ```bash
   python extralit/tests/integration/test_pymupdf_integration.py 
   python extralit/tests/integration/test_pymupdf_document_job.py 
   ```

4. Upload a PDF through the document upload job and verify:
   - Document upload completes successfully
   - PyMuPDF extraction is triggered
   - Markdown content and metadata are properly stored
   - Process continues gracefully even if hf-space is unavailable

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests have not been included_
- [ ] I need help with writing tests

**Tests added:**
- `extralit/tests/integration/test_pymupdf_integration.py` - Service health and basic extraction testing
- `extralit/tests/integration/test_pymupdf_document_job.py` - Full document workflow testing
- `extralit/tests/integration/PYMUPDF_TESTING_GUIDE.md` - Comprehensive testing documentation

## Added/updated documentations?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests have not been included_
- [ ] I need help with writing docs

**Documentation added:**
- Comprehensive testing guide with step-by-step instructions
- Function docstrings for new HTTP client service
- Integration workflow documentation

## Checklist
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

## Technical Details

**Files Modified:**
- `extralit-server/src/extralit_server/services/hf_space.py` - New HTTP client service
- `extralit-server/src/extralit_server/jobs/document_jobs.py` - Document job integration
- `extralit/tests/integration/test_pymupdf_*.py` - Integration tests

**Key Features:**
- ✅ HTTP client with proper timeout and error handling
- ✅ Graceful degradation when hf-space service unavailable
- ✅ Metadata serialization and merging
- ✅ Comprehensive test coverage with real PDF processing
- ✅ Environment-based configuration

priyankeshh committed:
feat: Add HTTP client for extralit-hf-space integration

- Create services/hf_space.py with HfSpaceClient for PyMuPDF extraction
- Add configurable settings with HF_SPACE_ environment variables
- Implement extract_pdf_with_pymupdf convenience function
- Add proper error handling and health check functionality
- Support analysis_metadata passing to hf-space service
- Use httpx for async HTTP requests with proper timeouts

priyankeshh committed:
feat: Integrate PyMuPDF extraction in document upload job

- Add import for extract_pdf_with_pymupdf from hf_space service
- Integrate PyMuPDF extraction after preprocessing step
- Store extraction results (markdown + metadata) in file metadata
- Add comprehensive error handling for extraction failures
- Continue with file upload even if extraction fails (non-blocking)
- Log extraction progress and results for monitoring

priyankeshh committed:
feat: Add integration test script for hf-space service

- Create comprehensive test script to verify hf-space integration
- Test HfSpaceClient creation and configuration
- Test health check endpoint connectivity
- Test PDF extraction with sample PDFs from benchmark directory
- Include detailed logging and error handling
- Provide clear success/failure indicators for debugging

priyankeshh committed:
feat: Add working integration test - successfully extracts 71k chars from 17-page PDF

priyankeshh committed:
feat: Fix PDFMetadata serialization and add end-to-end document job test

- Fixed JSON serialization issue with PDFMetadata objects
- Added comprehensive document upload job test
- Verified full workflow: preprocessing -> PyMuPDF extraction -> metadata merge
- Successfully extracts 71k chars and merges all metadata fields

priyankeshh committed:
docs: Organize test files and add comprehensive testing guide

- Moved integration tests to proper integration/ folder
- Renamed tests to follow existing naming convention
- Added detailed PYMUPDF_TESTING_GUIDE.md with step-by-step instructions
- Tests cover: health checks, PDF extraction, document job workflow
- Ready for manual testing and validation

JonnyTran reviewed (changes requested):
I think it'll be a good idea to move the `ExtractionRequest` and `ExtractionResponse` to the same file where you have the `HfSpaceClient` or to `extralit-server/src/extralit_server/api/schemas/v1/document/ocr.py`, where the extralit-hf-space will also import.

This will keep the API definition and implementation interface in the primary codebase.

priyankeshh committed:
feat: Add centralized OCR schemas in extralit-server

- Created extralit_server/api/schemas/v1/document/ocr.py with all OCR schemas
- Moved ExtractionRequest, ExtractionResponse, PyMuPDFExtractionResult to central location
- Single source of truth for schema definitions
- Improves code organization and maintainability

priyankeshh committed:
perf: Optimize HTTP client with connection reuse and centralized schemas

- Implement singleton pattern for HfSpaceClient to avoid initialization overhead
- Add global client management with get_client() function
- Remove duplicate PyMuPDFExtractionResult model, use centralized schema
- Add cleanup_client() for proper resource management
- Improve error handling and logging consistency
- Reduces connection overhead for high-throughput PDF processing

priyankeshh committed:
refactor: Reorganize PyMuPDF client to contexts/ocr/text.py

- Move HTTP client from services/hf_space.py to contexts/ocr/text.py for better organization
- Rename classes for clarity: HfSpaceClient -> PyMuPDFExtractionClient
- Rename functions: extract_pdf_with_pymupdf -> extract_pdf_text
- Update all imports in document jobs and test files
- Add proper module exports in ocr/__init__.py
- Environment variables changed from HF_SPACE_* to PYMUPDF_*
- Improves code organization and follows existing architecture patterns

priyankeshh committed:
Merge develop into feat/pymupdf-server-integration

- Resolved conflicts in document_jobs.py by keeping PyMuPDF import alongside existing imports
- Resolved conflicts in ocr.py by adding Apache license header to PyMuPDF schemas
- All PyMuPDF integration functionality preserved during merge

priyankeshh committed:
Complete PyMuPDF integration improvements

- Phase 1: Schema consolidation with centralized definitions
- Phase 2: Performance optimization with connection reuse and singleton pattern
- Phase 3: File organization with descriptive naming (hf_space.py  text.py)
- Phase 4: Performance testing with comprehensive benchmarks
- Phase 5: Test updates with simplified parameters

Performance metrics:
 Processes 0.3-0.4MB, 1-2 page research papers in ~8-13 seconds
 Extracts 16,915-68,344+ characters of structured markdown
 Handles concurrent requests efficiently (5 concurrent connections)
 Memory efficient with proper cleanup
 100% success rate across all test scenarios

priyankeshh committed:
Add PyMuPDF performance testing tools

- Added comprehensive performance test script equivalent to ab -n 5 -c 5
- Tests multiple PDFs with detailed metrics and Apache Bench style output
- Generates documentation-ready performance statistics
- Interactive mode for single PDF or comprehensive testing

JonnyTran commented on extralit-server/src/extralit_server/contexts/ocr/text.py:
```diff
@@ -12,3 +12,236 @@
+
+
+# Public API exports
+__all__ = [
```
This is not needed

priyankeshh committed:
removed duplicate classes

priyankeshh committed:
Apply pre-commit formatting and linting fixes

priyankeshh committed:
clean up ocr module structure

- remove PyMuPDFExtractionResult duplicate class
- switch to direct imports instead of package-level exports
- remove __all__ declarations - not needed anymore
- cleaner and easier to maintain
